diff --git a/include/__config b/include/__config
index bccf90d1d..6630b7138 100644
--- a/include/__config
+++ b/include/__config
@@ -814,6 +814,7 @@ typedef __char32_t char32_t;
         defined(__GNU__) ||                                                                                            \
         defined(__APPLE__) ||                                                                                          \
         defined(__MVS__) ||                                                                                            \
+        defined(__OHOS_FAMILY__) || \
         defined(_AIX) ||                                                                                               \
         defined(__EMSCRIPTEN__)
 // clang-format on
@@ -888,7 +889,7 @@ typedef __char32_t char32_t;
 #  endif
 
 #  if defined(__BIONIC__) || defined(__NuttX__) || defined(__Fuchsia__) || defined(__wasi__) ||                        \
-      defined(_LIBCPP_HAS_MUSL_LIBC) || defined(__OpenBSD__) || defined(__LLVM_LIBC__)
+      defined(_LIBCPP_HAS_MUSL_LIBC) || defined(__OpenBSD__) || defined(__LLVM_LIBC__) || defined(__OHOS_FAMILY__)
 #    define _LIBCPP_PROVIDES_DEFAULT_RUNE_TABLE
 #  endif
 
diff --git a/include/__locale_dir/locale_base_api.h b/include/__locale_dir/locale_base_api.h
index 8c000c558..48be6475c 100644
--- a/include/__locale_dir/locale_base_api.h
+++ b/include/__locale_dir/locale_base_api.h
@@ -25,6 +25,8 @@
 #  include <__locale_dir/locale_base_api/fuchsia.h>
 #elif defined(__wasi__) || defined(_LIBCPP_HAS_MUSL_LIBC)
 #  include <__locale_dir/locale_base_api/musl.h>
+#elif defined(__LITEOS__) || defined(__OHOS__)
+#  include <__locale_dir/locale_base_api/musl.h>
 #elif defined(__APPLE__) || defined(__FreeBSD__)
 #  include <xlocale.h>
 #endif
